app/controllers/openid_application_controller.rb

```diff
@@ -37,29 +37,37 @@ def authenticate_token
 
     # Only want to fetch the Okta profile if the session isn't already established and not a CC token
     @session = Session.find(token) unless token.client_credentials_token?
-    profile = fetch_profile(token.identifiers.okta_uid) unless token.client_credentials_token? || !@session.nil?
-    populate_ssoi_token_payload(profile) if @session.nil? && !profile.nil? && profile.attrs['last_login_type'] == 'ssoi'
+    profile = @session.profile unless @session.nil? || @session.profile.nil?
+    profile = fetch_profile(token.identifiers.okta_uid) unless token.client_credentials_token? || !profile.nil?
+    populate_ssoi_token_payload(profile) if !profile.nil? && profile.attrs['last_login_type'] == 'ssoi'
+
+    if @session.nil? && !token.client_credentials_token?
+      establish_session(profile)
+      return false if @session.nil?
+    end
 
     # issued for a client vs a user
     if token.client_credentials_token? || token.ssoi_token?
-      if token.payload['scp'].include?('launch/patient')
-        launch = fetch_smart_launch_context
-        token.payload[:icn] = launch
-        token.payload[:launch] = { patient: launch } unless launch.nil?
-      end
-      if token.payload['scp'].include?('launch')
-        launch = fetch_smart_launch_context
-        token.payload[:launch] = base64_json?(launch) ? JSON.parse(Base64.decode64(launch)) : { patient: launch }
-      end
+      populate_payload_for_launch_patient_scope if token.payload['scp'].include?('launch/patient')
+      populate_payload_for_launch_scope if token.payload['scp'].include?('launch')
+
       return true
     end
 
-    establish_session(profile) if @session.nil?
-    return false if @session.nil?
-
     @current_user = OpenidUser.find(@session.uuid)
   end
 
+  def populate_payload_for_launch_scope
+    launch = fetch_smart_launch_context
+    token.payload[:launch] = base64_json?(launch) ? JSON.parse(Base64.decode64(launch)) : { patient: launch }
+  end
+
+  def populate_payload_for_launch_patient_scope
+    launch = fetch_smart_launch_context
+    token.payload[:icn] = launch
+    token.payload[:launch] = { patient: launch } unless launch.nil?
+  end
+
   def populate_ssoi_token_payload(profile)
     token.payload['last_login_type'] = 'ssoi'
     token.payload['icn'] = profile.attrs['icn']
```

Comment:
Semi-related to this PR, I think it would be good to get the sec id from the profile, rather than the token.payload['sub'].  https://github.com/department-of-veterans-affairs/vets-api/blob/master/modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb#L52

`token.payload['sec_id'] = profile.attrs['SecID']`